scrapy/core/scheduler.py

```diff
@@ -48,33 +153,56 @@ def __init__(self, dupefilter, jobdir=None, dqclass=None, mqclass=None,
 
     @classmethod
     def from_crawler(cls, crawler):
-        settings = crawler.settings
-        dupefilter_cls = load_object(settings['DUPEFILTER_CLASS'])
-        dupefilter = create_instance(dupefilter_cls, settings, crawler)
-        pqclass = load_object(settings['SCHEDULER_PRIORITY_QUEUE'])
-        dqclass = load_object(settings['SCHEDULER_DISK_QUEUE'])
-        mqclass = load_object(settings['SCHEDULER_MEMORY_QUEUE'])
-        logunser = settings.getbool('SCHEDULER_DEBUG')
-        return cls(dupefilter, jobdir=job_dir(settings), logunser=logunser,
-                   stats=crawler.stats, pqclass=pqclass, dqclass=dqclass,
-                   mqclass=mqclass, crawler=crawler)
-
-    def has_pending_requests(self):
+        """
+        Factory method, initializes the scheduler with arguments taken from the crawl settings
+        """
+        dupefilter_cls = load_object(crawler.settings['DUPEFILTER_CLASS'])
+        return cls(
+            dupefilter=create_instance(dupefilter_cls, crawler.settings, crawler),
+            jobdir=job_dir(crawler.settings),
+            dqclass=load_object(crawler.settings['SCHEDULER_DISK_QUEUE']),
+            mqclass=load_object(crawler.settings['SCHEDULER_MEMORY_QUEUE']),
+            logunser=crawler.settings.getbool('SCHEDULER_DEBUG'),
+            stats=crawler.stats,
+            pqclass=load_object(crawler.settings['SCHEDULER_PRIORITY_QUEUE']),
+            crawler=crawler,
+        )
+
+    def has_pending_requests(self) -> bool:
         return len(self) > 0
 
-    def open(self, spider):
+    def open(self, spider: Spider) -> Optional[Deferred]:
+        """
+        (1) initialize the memory queue
+        (2) initialize the disk queue if the ``jobdir`` attribute is a valid directory
+        (3) return the result of the dupefilter's ``open`` method
+        """
         self.spider = spider
         self.mqs = self._mq()
         self.dqs = self._dq() if self.dqdir else None
         return self.df.open()
 
-    def close(self, reason):
-        if self.dqs:
+    def close(self, reason: str) -> Optional[Deferred]:
+        """
+        (1) dump pending requests to disk if there is a disk queue
+        (2) return the result of the dupefilter's ``close`` method
+        """
+        if self.dqs is not None:
             state = self.dqs.close()
+            assert isinstance(self.dqdir, str)
```

Comment:
This is only to avoid a typing error in the next line (`_write_dqs_state` expects a `str` but it gets `Optional[str]`). At this point, if `self.dqs is not None` it's only because `self.dqdir` is a `str`.